feat: DAH-2280 return translations from listing service

cade-exygy opened the pull request:
## Description

Here is a demo video: https://www.loom.com/share/b6a0a8b2e0ab43c3a4c4cf32e9193aec?sid=332789a7-6f57-4a59-95e2-a0db5f3c4204

Listing Service now returns cached translations.

#Jira Ticket
DAH-2280

## Checklist before requesting review

### Version Control

- [x] branch name begins with `angular` if it contains updates to Angular code
- [x] branch name contains the Jira ticket number
- [x] PR name follows `type: TICKET-NUMBER Description` format, e.g. `feat: DAH-123 New Feature`

### Code quality

- [x] [the set of changes is small](https://google.github.io/eng-practices/review/developer/small-cls.html#what-is-small)
- [x] [if the set of changes cannot be small, reviewers have been forewarned](https://google.github.io/eng-practices/review/developer/small-cls.html#cant)
- [x] code meets test coverage thresholds
- [x] code is properly formatted
- [x] code is linted
- [x] code irrelevant to the ticket is not modified e.g. changing indentation due to automated formatting
- [x] automated tests pass
- [x] if the code changes the UI, it matches the UI design exactly

### Review instructions

- [x] instructions specify which environment(s) it applies to
- [x] instructions have already been performed at least once
- [x] instructions can be followed by PA testers
- [x] instructions specify if it can only be followed by an engineer

### Request review

- [x] PR has `needs review` label
- [x] Use `Housing Eng` group to automatically assign reviewers, and/or assign specific engineers
- [x] If time sensitive, notify engineers in Slack


Locally (Engineers only since we don't run the subscriber on start):
Make sure you have the env variables set:
GOOGLE_PROJECT_ID
GOOGLE_TRANSLATE_KEY

- run memcached
- run `yarn start`
- run `bundle exec rake translate_test:subscribe`
- run `tail -f log/development.log`
- Update a listing in salesforce, edit Credit History if you want to see the value appear on the listing details page. Credit History is just the field I used as example to ease testing. See #2249 
- You should receive an event and see those values get translated in your logs
- Go to that listing and view the network object. You should see those updated translations
- If you scroll down to Credit History, you should see that updated value.

cade-exygy commented on app/javascript/modules/listingDetails/ListingDetailsEligibility.tsx:
```diff
@@ -380,7 +385,13 @@ export const ListingDetailsEligibility = ({
-                  {stripMostTags(listing.Credit_Rating)}
+                  {stripMostTags(
+                    getTranslatedString(
+                      listing?.Credit_Rating,
```
This field is coming from the Salesforce Cache, which is not refreshed at the same time as the translations

cade-exygy replied:
Will be addressed in https://sfgovdt.jira.com/browse/DAH-2575

cade-exygy commented on app/javascript/pages/listings/listing-detail.tsx:
```diff
@@ -53,6 +53,7 @@ const ListingDetail = () => {
   const isApplicationOpen = listing && isOpen(listing)
   const listingIsHabitat = listing && isHabitatListing(listing)
+  // TODO: Remove this function call once Google Cloud Translate is fully implemented
   useTranslate()
```
I would recommend commenting this line out for testing

cade-exygy commented on app/services/force/event_subscriber_translate_service.rb:
```diff
@@ -19,6 +19,10 @@ class EventSubscriberTranslateService
     def initialize
       setup_salesforce_client
       setup_faye_client
+      @translation_service = GoogleTranslationService.new(
```
Not sure if we want to / need to initialize GoogleTranslationService here.

jimlin-sfgov replied:
Yes, it would be best to follow the initialization pattern of other services in this file.

jimlin-sfgov commented:
@cade-exygy This is feedback for the PR description/template, not the code itself.
- write out one to two sentences for the Description section, the demo link is good to keep
- paste the Jira ticket link underneath the `# Jira ticket` title section, instead of replacing text
- keep review instructions in the Jira ticket as before, but if you want to have a copy of it in Github, paste it at the bottom of the PR description, instead of in the middle of the checklist

tallulahkay commented on app/javascript/modules/listingDetails/ListingDetailsEligibility.tsx:
```diff
@@ -111,6 +112,10 @@ export const ListingDetailsEligibility = ({
     }
   })
 
+  console.log(
```
extra log?

tallulahkay commented on app/services/force/event_subscriber_translate_service.rb:
```diff
@@ -108,9 +113,15 @@ def parse_event(platform_event)
+
+      logger(
+        "No updated values found in Salesforce event: #{event.inspect}",
+        error: true,
```
does this need to be an error?

jimlin-sfgov commented:
@cade-exygy feedback about the review instructions:
1. the last bullet mentions the Credit History field for the first time, are reviewers suppose to modify that field specifically in an earlier?
2. why just the Credit History the field?
3. the instructions appear to only be feasible locally, and by an engineer. please follow up on these checklist items:
    - 'instructions can be followed by PA testers'
    - 'instructions specify if it can only be followed by an engineer'

jimlin-sfgov commented on app/services/force/event_subscriber_translate_service.rb:
```diff
@@ -108,9 +113,15 @@ def parse_event(platform_event)
 
     def extract_updated_values(changed_fields, event)
-      changed_fields.each_with_object({}) do |field, values|
+      updated_values = changed_fields.each_with_object({}) do |field, values|
```
question: Do we still need to mutate `changed_fields`? If not we could use `map` instead:
```ruby
updated_values = changed_fields
                 .keys
                 .except('LastModifiedDate')
                 .map { |field| event.dig('payload', field) }
```

jimlin-sfgov commented on app/javascript/api/types/rails/listings/RailsTranslation.ts:
```diff
@@ -0,0 +1,15 @@
+  ES = "ES",
+  ZH = "ZH",
+  TL = "TL",
+}
```
nit: Add a space between declarations. Also, strange that our linter/formatter doesn't catch this.

cade-exygy commented:
@tallulahkay Can you review this again? I cleaned up a few things and addressed some of the bugs we have encountered. I also added the feature toggle but only for the React side. We will need to integrate Unleash SDK on the Rails side to include feature toggles there. 

I also added the Google Translate API key to the deplolyed heroku instance if you want to test there. (You still will have to run memcached and start the subscriber service.)

Let me know if you catch anything!

tallulahkay commented on app/services/cache_service.rb:
```diff
@@ -61,7 +58,7 @@ def process_translations(listing)
     translation_service.cache_listing_translations(
       listing['Id'],
-      strings_to_translate.keys,
+      strings_to_translate.keys.map { |key| "#{key}__c" },
```
I don't actually think we need to do this - you can just leave the key as is, then when you're calling it in `getTranslatedString` it can be:

getTranslatedString(
  listing,
  Reserved_community_type_Description,
)

tallulahkay commented on app/helpers/service_helper.rb:
```diff
@@ -0,0 +1,48 @@
+module ServiceHelper
+  SALESFORCE_LISTING_FIELD_NAMES_TO_TRANSLATE = %w[
```
So that we don't have to keep two constants in sync, can we just use LISTING_FIELDS_TO_TRANSLATE and append the __c for when this is used?

tallulahkay commented on app/services/google_translation_service.rb:
```diff
@@ -62,4 +68,12 @@ def transform_translations_for_caching(listing_id, keys, translations)
     return_value
   end
+
+  def logger(message, error: false)
```
Can we call this google_translation_error or something similar? So that it makes it clear when it's being called that we are calling a specific error handler for this service

tallulahkay commented on app/javascript/pages/listings/listing-detail.tsx:
```diff
@@ -37,25 +37,27 @@ import { MobileListingDetailsLottery } from "../../modules/listingDetailsLottery
 
 const ListingDetail = () => {
+  const isCloudTranslationEnabled = useFeatureFlag("GoogleCloudTranslate", true)
+  useTranslate(isCloudTranslationEnabled)
```
why not call the feature flag inside of the useTranslate hook?

tallulahkay commented on app/javascript/util/languageUtil.tsx:
```diff
@@ -227,3 +231,31 @@ export function localizedFormat(date: string, format: string): string {
+  translations: RailsTranslations
+) => {
+  const languageCode = getPageLanguageCode()
+  const translatedValue = translations ? (translations[fieldName]?.[languageCode] as string) : null
```
you need like an isEmpty(translations) instead of just translations, because translations is coming in as an empty object and that is true for the purposes of the ternary

tallulahkay reviewed (changes requested):
Works well, just a few requests for code changes!

I'm worried about the diff issue from salesforce - has Reggie gotten back to you about that?

cade-exygy commented on app/javascript/__tests__/hooks/useTranslate.test.tsx:
```diff
@@ -0,0 +1,62 @@
+    expect(dispatchEventSpy).toHaveBeenCalled()
+  })
+
+  it("render dom element for translation when disabled", () => {
```
change title

cade-exygy commented on app/javascript/hooks/useTranslate.tsx:
```diff
@@ -37,8 +37,8 @@ const initGoogleTranslate = () => {
 
-const useTranslate = (disable) => {
-  const languageInRoute = getCurrentLanguage()
+const useTranslate = (disable: boolean = true) => {
```
if the feature flag value isn't ready (undefined), then we default `disable` to true

tallulahkay reviewed (approved):
Thanks for all the hard work!
